fuel-vm/src/interpreter/flow.rs

```diff
@@ -149,138 +161,385 @@ where
+            *self.registers.read_registers.cgas,
+            self.params.amount_of_gas_to_forward,
+        );
 
         // subtract gas
-        self.registers[RegId::CGAS] = arith::sub_word(self.registers[RegId::CGAS], forward_gas_amount)?;
+        *self.registers.read_registers.cgas = arith::sub_word(*self.registers.read_registers.cgas, forward_gas_amount)?;
 
-        *frame.set_context_gas() = self.registers[RegId::CGAS];
-        *frame.set_global_gas() = self.registers[RegId::GGAS];
+        *frame.set_context_gas() = *self.registers.read_registers.cgas;
+        *frame.set_global_gas() = *self.registers.read_registers.ggas;
 
         let frame_bytes = frame.to_bytes();
         let len = arith::add_word(frame_bytes.len() as Word, frame.total_code_size())?;
 
-        if len > self.registers[RegId::HP] || self.registers[RegId::SP] > self.registers[RegId::HP] - len {
+        if len > *self.registers.read_registers.hp
+            || *self.registers.read_registers.sp > *self.registers.read_registers.hp - len
+        {
             return Err(PanicReason::MemoryOverflow.into());
         }
+        let id = internal_contract_or_default(
+            self.context,
+            self.registers.read_registers.fp.as_ref(),
+            self.memory.memory,
+        );
 
-        let id = self.internal_contract_or_default();
-
-        self.set_frame_pointer(self.registers[RegId::SP]);
-
-        self.registers[RegId::SP] += len;
-        self.registers[RegId::SSP] = self.registers[RegId::SP];
-
-        let fpx = arith::add_word(self.registers[RegId::FP], frame_bytes.len() as Word)?;
-        self.memory[self.registers[RegId::FP] as usize..fpx as usize].copy_from_slice(frame_bytes.as_slice());
+        let sp = *self.registers.read_registers.sp;
+        set_frame_pointer(self.context, self.registers.read_registers.fp.as_mut(), sp);
 
-        let code_range = (fpx as usize)..arith::add_usize(fpx as usize, frame.code_size() as usize);
-        let bytes_read = self
-            .storage
-            .read(frame.to(), &mut self.memory[code_range.clone()])
-            .map_err(RuntimeError::from_io)?
-            .ok_or(PanicReason::ContractNotFound)?;
-        if bytes_read as Word != frame.code_size() {
-            return Err(PanicReason::ContractNotFound.into());
-        }
-        let pad_len = frame.code_size_padding();
-        if pad_len > 0 {
-            self.memory[code_range.end..code_range.end + pad_len as usize]
-                .copy_from_slice(&[0; 32][..pad_len as usize]);
-        }
+        *self.registers.read_registers.sp += len;
+        *self.registers.read_registers.ssp = *self.registers.read_registers.sp;
 
-        self.registers[RegId::BAL] = b;
-        self.registers[RegId::PC] = fpx;
-        self.registers[RegId::IS] = self.registers[RegId::PC];
-        self.registers[RegId::CGAS] = forward_gas_amount;
+        let code_mem_range = CheckedMemRange::new(*self.registers.read_registers.fp, len as usize)?;
```

Comment:
[nit] this range includes more than just code

```suggestion
        let frame_mem_range = CheckedMemRange::new(*self.registers.read_registers.fp, len as usize)?;
```